### PR TITLE
ASoC: hdac_hda: fix memleak on module unload

### PR DESCRIPTION
**Pull request by kv2019i**
Set of patches to use common cleanup code to free hda codec resources.

Fixes: https://github.com/thesofproject/linux/issues/2195

### CODE REVIEW

File: sound/soc/codecs/hdac_hda.c [hunk 1]
```diff
@@ -603,12 +607,7 @@ static int hdac_hda_dev_probe(struct hdac_device *hdev)
 
 static int hdac_hda_dev_remove(struct hdac_device *hdev)
 {
-	struct hdac_hda_priv *hda_pvt;
-
-	hda_pvt = dev_get_drvdata(&hdev->dev);
-	if (hda_pvt && hda_pvt->codec.registered)
-		cancel_delayed_work_sync(&hda_pvt->codec.jackpoll_work);
-
+	/* empty function needed for hda_codec_driver_remove() */
```
Comment (plbossart) [inline sound/soc/codecs/hdac_hda.c]: I wasn't able to follow the commit message, the comment above and where the cancel delayed work would be done now?
@kv2019i maybe a couple of explanations would help here?

**Comment by aiChaoSONG** [comment]
SOFCI TEST

File: sound/soc/codecs/hdac_hda.c [hunk 2]
```diff
@@ -519,7 +520,10 @@ static void hdac_hda_codec_remove(struct snd_soc_component *component)
 	pm_runtime_disable(&hdev->dev);
 	snd_hdac_ext_bus_link_put(hdev->bus, hlink);
 
-	snd_hdac_regmap_exit(hdev);
+	if (codec->patch_ops.free)
+		codec->patch_ops.free(codec);
+
+	snd_hda_codec_cleanup_for_unbind(codec);
```
Comment (lyakh) [inline sound/soc/codecs/hdac_hda.c:526]: wow, this looks like a lot of clean up was missing? I see `snd_hda_codec_cleanup_for_unbind()` is currently called on the clean up path from `hda_codec_driver_remove()`. Is that function not called at all for this codec in the mem-leak case? Should it be called? Just for my information: how are these two clean up routines (codec and component `.remove()`) supposed to work together in this case?

Comment (kv2019i) [reply sound/soc/codecs/hdac_hda.c:526]: @lyakh yeah, maybe I should use stronger wording in the commit message -- we missed releasing 95% of allocated resources. I have looked at the cleanup routines many times, but it never occured to me to check that they are actually called... and it appears they are not. This is not so easy to pick-up with code review. hda_codec_driver_remove() is called but as ASoC/SOF sets the hdac-ext ops, the removal patch just calls "return codec->bus->core.ext_ops->hdev_detach(&codec->core);"  For the latter question, component remove is called when the ASoC card is removed and codec remote is called later when the codec driver module is finally unloaded. The HDA codec remove functions have some bits that assume the parent card to be still around, so we have to do most of the the cleanup in component remove.

File: sound/soc/codecs/hdac_hda.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/codecs/hdac_hda.c]: Ack @plbossart . I'll do another round and add a bit more comments. This is somewhat hard to follow with the extended HDA bus and ASoC, laid on top of bare HDA interfaces.

File: sound/soc/codecs/hdac_hda.c [hunk 3]
```diff
@@ -603,12 +607,10 @@ static int hdac_hda_dev_probe(struct hdac_device *hdev)
...
 static int hdac_hda_dev_remove(struct hdac_device *hdev)
 {
-	struct hdac_hda_priv *hda_pvt;
-
-	hda_pvt = dev_get_drvdata(&hdev->dev);
-	if (hda_pvt && hda_pvt->codec.registered)
-		cancel_delayed_work_sync(&hda_pvt->codec.jackpoll_work);
-
+	/*
+	 * Resources are freed hdac_hda_codec_remove(). This
```
Comment (ranj063) [inline sound/soc/codecs/hdac_hda.c:611]: only minor , a missing "in"

Comment (plbossart) [reply sound/soc/codecs/hdac_hda.c:611]: @kv2019i can you add a fixup! PR for this and merge it directly? no point in everyone reviewing this.

Comment (kv2019i) [reply sound/soc/codecs/hdac_hda.c:611]: Doh, sorry @ranj063 and @plbossart , I missed this. I'll make a fixup.

Comment (kv2019i) [reply sound/soc/codecs/hdac_hda.c:611]: Tackled in https://github.com/thesofproject/linux/pull/2262
